feat(eslint-config-react-app): Add jest & testing-library rules

MichaelDeBoey opened the pull request:
Closes #8446
Closes #8959

Since we now have ESLint 7 support, we need to wait until `eslint-plugin-jest-dom` & `eslint-plugin-testing-library` are ESLint 7 compatible.

- [x] [`eslint-plugin-jest`](https://github.com/jest-community/eslint-plugin-jest/issues/568)
  - [x]  [PR](https://github.com/jest-community/eslint-plugin-jest/pull/572)
  - [x]  Released
- [x] [`eslint-plugin-testing-library`](https://github.com/testing-library/eslint-plugin-testing-library/issues/138)
  - [x]  [PR](https://github.com/testing-library/eslint-plugin-testing-library/pull/139)
  - [x]  [Released](https://github.com/testing-library/eslint-plugin-testing-library/releases/tag/v3.9.0)

eddiemonge commented on packages/eslint-config-react-app/index.js:
```diff
@@ -22,13 +22,39 @@
+  const allDeps = Object.keys({
+    ...pkg.peerDependencies,
+    ...pkg.devDependencies,
+    ...pkg.dependencies,
```
are there any optionalDependencies? or does that not matter

MichaelDeBoey replied:
Doesn't matter here I think

eddiemonge commented on packages/eslint-config-react-app/index.js:
```diff
@@ -22,13 +22,39 @@
+    'jsx-a11y',
+    'react',
+    'react-hooks',
+    hasJestDom ? 'jest-dom' : null,
```
```suggestion
    hasJestDom && 'jest-dom',
```

MichaelDeBoey replied:
Any reason why this is recommended? 🤔

eddiemonge replied:
less code. accomplishes the same thing

MichaelDeBoey replied:
Using a ternary is how it's done in `kcd-scripts` and is how it's recommended by @kentcdodds itself, so going to leave it like this

eddiemonge commented on packages/eslint-config-react-app/index.js:
```diff
@@ -22,13 +22,39 @@
+    'react',
+    'react-hooks',
+    hasJestDom ? 'jest-dom' : null,
+    hasTestingLibrary ? 'testing-library' : null,
```
```suggestion
    hasTestingLibrary && 'testing-library',
```

eddiemonge commented:
Dupe of #8155 ?

MichaelDeBoey commented:
Ping @iansu

NMinhNguyen commented on packages/eslint-config-react-app/index.js:
```diff
@@ -315,5 +341,37 @@ module.exports = {
     'flowtype/use-flow-type': 'warn',
+
+    // https://github.com/testing-library/eslint-plugin-jest-dom
+    ...(hasJestDom
```
FYI the original PR only adds these rule to test files which is probably better: https://github.com/facebook/create-react-app/pull/8155/files#diff-b4a510397b084f197f379c4e3485cde1

ianschmitz commented:
I have the same question as here: https://github.com/facebook/create-react-app/pull/8155#issuecomment-565302119

NMinhNguyen commented:
@ianschmitz I've replied on the original PR https://github.com/facebook/create-react-app/pull/8155#issuecomment-647026881

nickmccurdy commented:
I'd really appreciate it if we can merge this or #8155. It's important to the user experience of testing library users that the ESLint plugins are used, but it's confusing that CRA includes some of our best practice suggestions while missing the ESLint plugins that make it easy for users to follow them. I understand not having this in the build tool is not ideal, but this seems to be an issue with `jest-runner-eslint` that hasn't been fixed for months, and most users are in an editor anyway and not always looking at the CLI output.

MichaelDeBoey commented:
@nickmccurdy @ianschmitz This one can be merged if it's up to me. 🙂

ianschmitz commented on packages/eslint-config-react-app/index.js:
```diff
@@ -315,5 +341,37 @@ module.exports = {
     'flowtype/use-flow-type': 'warn',
+
+    // https://github.com/testing-library/eslint-plugin-jest-dom
+    ...(hasJestDom
```
Is there harm in always including jest-dom rules even if using the node test environment?

kentcdodds replied:
It should be fine 👍

ianschmitz commented on packages/eslint-config-react-app/index.js:
```diff
@@ -315,5 +341,37 @@ module.exports = {
+      : null),
+
+    // https://github.com/testing-library/eslint-plugin-testing-library
+    ...(hasTestingLibrary
```
Similarly if these are specific to testing-library APIs, could we always enable and avoid the dependency check? Any harm?

MichaelDeBoey commented:
@ianschmitz I've rebased my branch onto `upstream/master` and removed the check.
Since we now have ESLint 7 support, we need to wait until `eslint-plugin-jest-dom` & `eslint-plugin-testing-library` are ESLint 7 compatible.

- [x] [`eslint-plugin-jest-dom`](https://github.com/testing-library/eslint-plugin-jest-dom/issues/45)
  - [x]  [PR](https://github.com/testing-library/eslint-plugin-jest-dom/pull/46)
  - [x]  [Release](https://github.com/testing-library/eslint-plugin-jest-dom/releases/tag/v3.0.0)
- [ ] [`eslint-plugin-testing-library`](https://github.com/testing-library/eslint-plugin-testing-library/issues/138)
  - [ ]  [PR](https://github.com/testing-library/eslint-plugin-testing-library/pull/139)
  - [ ]  Released

MichaelDeBoey commented:
@ianschmitz If needed, I can add `eslint-plugin-jest` too like it's done in #8155.

nickmccurdy commented:
I've removed the plugin from #8155 because it's still error (even at the latest version), but if you can figure out how to fix it that would be great https://github.com/nickmccurdy/create-react-app/tree/testing-library-linting-integration

MichaelDeBoey commented:
@nickmccurdy Then this one can be merged as soon as `eslint-plugin-testing-library` is ESLint 7 compatible 🙂

nickmccurdy commented:
Oh sorry I misread, i thought you were talking about `jest-runner-eslint`, ignore me.

eddiemonge commented:
Sadly this will never get merged if its waiting on `eslint-plugin-testing-library` because of [circular dependencies](https://github.com/mysticatea/eslint-plugin/pull/29#issuecomment-651908435). I'm hoping somewhere in the chain is a typo

Belco90 commented:
I missed the last part of the discussion. What does eslint-plugin-testing-library to wait for ESLint v7 to get merged?

Apart from that, @michaeldeboey created a PR in eslint-plugin-testing-library for making it compatible with ESLint v7. We are waiting some dependencies there to be bumped too, but our tests are passing with v7 already so I think none of the breaking changes affect us and we can merge that PR safely.

I’m missing something else I guess.

MichaelDeBoey commented:
@eddiemonge The circular dependency will be resolved once `eslint-utils` is merged without updating `@mysticatea/eslint-plugin`, like I explained in https://github.com/mysticatea/eslint-utils/pull/7#issuecomment-651940524 🙂

nickmccurdy commented on packages/eslint-config-react-app/index.js:
```diff
@@ -315,5 +323,30 @@ module.exports = {
     'flowtype/use-flow-type': 'warn',
+
+    // https://github.com/testing-library/eslint-plugin-jest-dom
+    'jest-dom/prefer-checked': 'error',
+    'jest-dom/prefer-empty': 'error',
+    'jest-dom/prefer-enabled-disabled': 'error',
+    'jest-dom/prefer-focus': 'error',
+    'jest-dom/prefer-required': 'error',
+    'jest-dom/prefer-to-have-attribute': 'error',
+    'jest-dom/prefer-to-have-text-content': 'error',
```
We could just use the recommended config instead

nickmccurdy commented on packages/eslint-config-react-app/index.js:
```diff
@@ -315,5 +323,30 @@ module.exports = {
     'flowtype/define-flow-type': 'warn',
     'flowtype/require-valid-file-annotation': 'warn',
     'flowtype/use-flow-type': 'warn',
+
+    // https://github.com/testing-library/eslint-plugin-jest-dom
+    'jest-dom/prefer-checked': 'error',
+    'jest-dom/prefer-empty': 'error',
+    'jest-dom/prefer-enabled-disabled': 'error',
+    'jest-dom/prefer-focus': 'error',
+    'jest-dom/prefer-required': 'error',
+    'jest-dom/prefer-to-have-attribute': 'error',
+    'jest-dom/prefer-to-have-text-content': 'error',
```
We could just use the recommended config instead

MichaelDeBoey replied:
We're always defining the rules ourselves, so that's why I went with it here too.

nickmccurdy commented on packages/eslint-config-react-app/index.js:
```diff
@@ -315,5 +323,30 @@ module.exports = {
+    'jest-dom/prefer-required': 'error',
+    'jest-dom/prefer-to-have-attribute': 'error',
+    'jest-dom/prefer-to-have-text-content': 'error',
+
+    // https://github.com/testing-library/eslint-plugin-testing-library
+    'testing-library/await-async-query': 'error',
+    'testing-library/await-async-utils': 'error',
+    'testing-library/await-fire-event': 'off',
+    'testing-library/consistent-data-testid': 'off',
+    'testing-library/no-await-sync-query': 'error',
+    'testing-library/no-debug': 'error',
+    'testing-library/no-dom-import': ['error', 'react'],
+    'testing-library/no-manual-cleanup': 'error',
+    'testing-library/no-wait-for-empty-callback': 'error',
+    'testing-library/prefer-explicit-assert': 'warn',
+    'testing-library/prefer-find-by': 'error',
+    'testing-library/prefer-presence-queries': 'error',
+    'testing-library/prefer-screen-queries': 'error',
+    'testing-library/prefer-wait-for': 'error',
```
Let's also use the recommended React config here so it's easier to compare the differences

MichaelDeBoey replied:
Same

kentcdodds commented:
If we're going to go with a recommended config, let's make sure that it's not too opinionated. One of the things I love about the `eslint-config-react-app` config is that even if a rule is (subjectively) "a good idea", it's not enabled, it really has to be an objectively good idea.

I'm not talking about any rules in particular. Just mentioning that here because there's nothing more annoying than getting a red underline when you don't understand why a change is necessary.

Ok, actually, here's an example. The new `no-multiple-assertions-wait-for` rule that was merged last week is a good idea, but I don't think that it's a good enough idea to strictly enforce in the default ruleset in `react-scripts`. The downside of multiple assertions does exist, but it's a relatively lightweight downside. Anyway, let's just keep this in mind generally :)

nickmccurdy commented:
Good point. While I'd personally want rules in my own projects to be as strict as possible to reduce tech debt, that doesn't always make sense for CRA users, especially if they're currently using older versions of testing library without some newer features.

Going along with the discussion in https://github.com/testing-library/eslint-plugin-testing-library/pull/192#issuecomment-652530685, what would you think about warning about rules like this by default rather than leaving them off? `create-react-app`'s default template already uses `<StrictMode>` by default, which logs warnings about features that are being removed from React or not recommended any more. I think using a warning here for things like new `user-event` features would be more consistent.

renatoagds commented:
One more thing that I would like to put on the table about `warn` and `error`, is to make sure that we won't set `error` for rules that is based on the last major version of RTL, like `testing-library/prefer-wait-for`. We have almost 4 months since RTL v10 and we probably have a lot of codebases that still in migration for v10 (myself included). I would like to keep up to date with latest version of `react-scripts` without getting a lot of errors.

nickmccurdy commented:
Yea, that's exactly what I think we should be warning for

MichaelDeBoey commented:
@ianschmitz @nickmccurdy @Belco90 @kentcdodds I've rebased onto `upstream/master` and made all preference rules `warn` instead of `error`;

MichaelDeBoey commented:
Should I add `eslint-plugin-jest` too btw? 🤔

Belco90 commented:
Well, not all of them are opinionated. Rules like `await-async-query` or `await-async-utils` are anticipating actual errors. In my opinion these kind of rules should be enabled as errores, and leave the opinionated ones as warnings (from React shareable config I mean).

MichaelDeBoey commented:
@Belco90 `await-async-query` & `await-async-utils` are still throwing errors

MichaelDeBoey commented:
I've added `eslint-config-jest` too 🙂

bennettdams commented:
While we wait: Should/could we prepare documentation on how to actually use the new configuration together with VS Code (and TypeScript & Prettier)?
I created a proposal some weeks ago:   https://github.com/facebook/create-react-app/issues/8849

MichaelDeBoey commented:
@ianschmitz Why is this a breaking change? 🤔

ianschmitz commented:
This PR enables new rules by default in `eslint-config-react-app`, which is used by many projects that aren't CRA.

NMinhNguyen commented:
Just wondering if there's a reason not to apply these rules only to test files like was done in the original PR https://github.com/facebook/create-react-app/pull/8155/files#diff-b4a510397b084f197f379c4e3485cde1? Would reduce the chance of false positives reported on source code? Perhaps that could be a reason to re-open the original PR and fix up any outstanding issues there?

nickmccurdy commented:
I think this PR is more up to date at this point, but you're welcome to borrow configuration from mine if you want.

mrmckeb commented:
@NMinhNguyen, I agree with you - but this is also safer. If those libraries have version bumps, that can introduce new rules by surprise.

mrmckeb commented:
@nickmccurdy I'm going to resolve the conflicts quickly and help this get merged in, but @ianschmitz pointed out that we're waiting on `eslint-plugin-testing-library`. Is this breaking something today?

mrmckeb commented:
@MichaelDeBoey I wanted to resolve the conflicts quickly and help this get merged in, but @ianschmitz pointed out that we're waiting on `eslint-plugin-testing-library`. Is this breaking something today? 

I've also noted some differences between these rules and their recommended configs. Is this intentional - or perhaps those configs have changed since you created this PR? 

For example, `eslint-plugin-jest` has `'jest/expect-expect'` set to `'warn'`, but here it's `'off'`. I've noted the same for `'jest/no-jasmine-globals'`, `'jest/no-standalone-expect'` and others.

NMinhNguyen commented:
> @NMinhNguyen, I agree with you - but this is also safer. If those libraries have version bumps, that can introduce new rules by surprise.

Ah I wasn't actually talking about using the recommended presets, more so about using `overrides` and  `'**/__tests__/**/*.{js,jsx,ts,tsx}'`. You can still specify `rules` like is done in this PR and not rely on `extends`:

```js
    {
      files: [
        '**/__tests__/**/*.{js,jsx,ts,tsx}',
        '**/*.{spec,test}.{js,jsx,ts,tsx}',
      ],
      rules: [
        'jest/no-large-snapshots': ['warn', { maxSize: 300 }],
        // ...
      ],
      plugins: [
        'jest-dom',
        'testing-library',
      ]
    },
```

MichaelDeBoey commented:
> I've also noted some differences between these rules and their recommended configs. Is this intentional - or perhaps those configs have changed since you created this PR?
> 
> For example, `eslint-plugin-jest` has `'jest/expect-expect'` set to `'warn'`, but here it's `'off'`. I've noted the same for `'jest/no-jasmine-globals'`, `'jest/no-standalone-expect'` and others.

@mrmckeb I think I took `kcd-scripts`' rules as a starting point.

mrmckeb commented:
Thanks @MichaelDeBoey - I might try to align these better with the recommended rules then.

Do we need to wait for this PR?
https://github.com/testing-library/eslint-plugin-testing-library/issues/138

mrmckeb commented:
Thanks for the clarification @NMinhNguyen, that makes a lot of sense. @ianmcnally and @iansu - do you have any thoughts on scoping the testing rules to test files? I'm 50/50 on this, but I think adding it now would be wise if we're unsure.

MichaelDeBoey commented:
> Do we need to wait for this PR?

@mrmckeb We're indeed still waiting for https://github.com/testing-library/eslint-plugin-testing-library/pull/139, but that one's still waiting for https://github.com/mysticatea/eslint-utils/pull/7 in the end.

@mysticatea was going to give it a look this weekend, but I guess he didn't have time unfortunately.

mrmckeb commented:
Thanks for the updates @MichaelDeBoey.

iansu commented:
@mrmckeb I do think scoping rules for RTL and Jest to test files only is a good idea.

MichaelDeBoey commented:
@ianschmitz @mrmckeb I rebased the branch onto `upstream/master`, so it has #9587.
I also changed the `jest` file instead of `index` so this PR is in line with that PR 🙂

Belco90 commented:
👀 https://github.com/testing-library/eslint-plugin-testing-library/releases/tag/v3.9.0

MichaelDeBoey commented:
@ianschmitz @mrmckeb I think we can merge this one now 🙂

mrmckeb commented:
This looks great @MichaelDeBoey, we just need to have a quick discussion on the final set of rules.

We generally don't include "prefer"-type rules, and stick to _potential errors_ only. I've invited you to the next meeting, so that might be a good place to go through these together?

It would be great if you could also include the source(s) for the rules as applied here.

We may also want to split the RTL rules out from the Jest rules, but I don't think we came to a conclusion on that.

MichaelDeBoey commented:
> We generally don't include "prefer"-type rules, and stick to _potential errors_ only.

@mrmckeb For `jest` I started by using @kentcdodds' `eslint-config-kentcdodds`.
For `jest-dom` & `testing-library` I took the rules Kent recommended (see https://github.com/facebook/create-react-app/issues/8959#issuecomment-624286066).

> I've invited you to the next meeting, so that might be a good place to go through these together?

We could do it earlier via DM if you want, so we can merge this one quicker.

> It would be great if you could also include the source(s) for the rules as applied here.

What do you exactly need from me for this?
I already added the link to all plugins, so I don't understand what you're trying to ask from me. 

> We may also want to split the RTL rules out from the Jest rules, but I don't think we came to a conclusion on that.

All rules are for testing, so I don't think RTL rules should be split out?

Belco90 commented:
Great news @mrchief! I'm sure @MichaelDeBoey can give you the info needed for setting up the plugin properly. Anyway, let me know if I can help with something else, specially the _potential errors_ rules only. Usually the "no-" type rules are the ones which apply here, but others as "await-async-query" or "await-async-utils" are preventing async errors too.

mrchief commented:
> Great news @mrchief! I'm sure @MichaelDeBoey can give you the info needed for setting up the plugin properly. Anyway, let me know if I can help with something else, specially the _potential errors_ rules only. Usually the "no-" type rules are the ones which apply here, but others as "await-async-query" or "await-async-utils" are preventing async errors too.

I think you meant to tag @mrmckeb ? 😀

Belco90 commented:
@mrchief Indeed! It was autocomplete's fault 😛

mrmckeb commented:
GitHub does that to me too @Belco90 - I'm not sure why it would prioritise people not in the current thread... 🤷  😆 

> I already added the link to all plugins, so I don't understand what you're trying to ask from me.
@MichaelDeBoey You've mentioned that you based this on Kent's config, so that's all we need for now I think. We'll probably want to strip a lot of those out and try to get to a smaller set of rules.

> All rules are for testing, so I don't think RTL rules should be split out?
It's important to keep in mind that Jest ships with `react-scripts`, but RTL is a template dependency. So we don't expect everyone to be using it (even if we recommend it). This is an open question for now though.

> I already added the link to all plugins, so I don't understand what you're trying to ask from me.
We could do a DM on Twitter, but we definitely need input from @ianschmitz and @iansu too ;)

MichaelDeBoey commented:
@mrmckeb So I've been over the rules and deleted the not recommended rules from `eslint-plugin-jest`

All `jest-dom` rules start with the 'prefer-' prefix, but are all recommended

Not all `testing-library` rules are recommended atm, but they're gonna be in V4 of the plugin

Belco90 commented:
> Not all testing-library rules are recommended atm, but they're gonna be in V4 of the plugin

What do you mean with this? In v4 we are updating recommended rules but not all are recommended.

Belco90 commented:
@MichaelDeBoey I guess you mean you enabled those rules which will be recommended in v4.

mrmckeb commented on packages/eslint-config-react-app/jest.js:
```diff
@@ -12,21 +12,62 @@
+        'jest/valid-expect-in-promise': 'error',
+        'jest/valid-title': 'warn',
+
+        // https://github.com/testing-library/eslint-plugin-jest-dom
+        'jest-dom/prefer-checked': 'warn',
+        'jest-dom/prefer-empty': 'warn',
+        'jest-dom/prefer-enabled-disabled': 'warn',
+        'jest-dom/prefer-focus': 'warn',
+        'jest-dom/prefer-required': 'warn',
+        'jest-dom/prefer-to-have-attribute': 'warn',
+        'jest-dom/prefer-to-have-style': 'warn',
+        'jest-dom/prefer-to-have-text-content': 'warn',
```
@ianschmitz and @iansu all rules for this plugin are `prefer`. Perhaps we should just remove it?

MichaelDeBoey replied:
All rules are recommended in the `jest-dom` plugin.

If not adhered there won't be any error, but I still think it's a good idea to at least warn the users and "teach" them how they should use `jest-dom`.

mrmckeb replied:
I understand sorry, it was more about aligning with the philosophy of the current ESLint config - which has always tried to avoid preferential rules.
https://github.com/facebook/create-react-app/blob/master/packages/eslint-config-react-app/index.js

But again, let's see what the rest of the team think :) Let's not do anything until we get that feedback.

@Timer may also have some thoughts on this.

mrmckeb replied:
We'll remove this for now, and consider adding in later - perhaps as an optional/extended set of rules.

mrmckeb commented on packages/eslint-config-react-app/jest.js:
```diff
@@ -12,21 +12,62 @@
-        'jest/valid-describe': 'warn',
-        'jest/valid-expect': 'warn',
-        'jest/valid-expect-in-promise': 'warn',
+        // https://github.com/jest-community/eslint-plugin-jest
+        'jest/expect-expect': 'off',
+        'jest/no-commented-out-tests': 'warn',
+        'jest/no-conditional-expect': 'error',
+        'jest/no-deprecated-functions': 'warn',
+        'jest/no-disabled-tests': 'warn',
+        'jest/no-done-callback': 'error',
+        'jest/no-export': 'error',
+        'jest/no-focused-tests': 'warn',
+        'jest/no-identical-title': 'error',
+        'jest/no-interpolation-in-snapshots': 'error',
+        'jest/no-jasmine-globals': 'error',
+        'jest/no-jest-import': 'error',
+        'jest/no-mocks-import': 'error',
+        'jest/no-standalone-expect': 'off',
+        'jest/no-test-callback': 'off',
+        'jest/no-test-prefixes': 'warn',
+        'jest/valid-describe': 'error',
+        'jest/valid-expect': 'error',
+        'jest/valid-expect-in-promise': 'error',
+        'jest/valid-title': 'warn',
```
I think we can also turn off:
- `jest/no-commented-out-tests`
- `jest/no-deprecated-functions` (TypeScript 4.0 should also help here)
- `jest/no-disabled-tests`
- `jest/no-done-callback`
- `jest/no-export`
- `jest/no-focused-tests`
- `jest/no-test-prefixes`

I'm not sure about this, maybe @ianschmitz can chime in:
- `jest/no-jasmine-globals`

Also, I couldn't `no-test-callback`, has that been deprecated?

MichaelDeBoey replied:
For now I kept the `jest/no-jasmine-globals` rule
Let me know if it needs to be removed 🙂

mrmckeb commented on packages/eslint-config-react-app/jest.js:
```diff
@@ -12,21 +12,62 @@
+        'jest-dom/prefer-to-have-style': 'warn',
+        'jest-dom/prefer-to-have-text-content': 'warn',
+
+        // https://github.com/testing-library/eslint-plugin-testing-library
+        'testing-library/await-async-query': 'error',
+        'testing-library/await-async-utils': 'error',
+        'testing-library/no-await-sync-query': 'warn',
+        'testing-library/no-debug': 'warn',
+        'testing-library/no-dom-import': ['error', 'react'],
+        'testing-library/no-manual-cleanup': 'error',
+        'testing-library/no-render-in-setup': 'error',
+        'testing-library/no-wait-for-empty-callback': 'error',
+        'testing-library/no-wait-for-snapshot': 'error',
+        'testing-library/prefer-explicit-assert': 'warn',
+        'testing-library/prefer-find-by': 'warn',
+        'testing-library/prefer-presence-queries': 'warn',
+        'testing-library/prefer-screen-queries': 'warn',
+        'testing-library/prefer-wait-for': 'warn',
```
I think we can also turn off:
- `testing-library/no-debug`
- `testing-library/no-manual-cleanup`
- `testing-library/no-render-in-setup`
- `testing-library/prefer-*`

I'm not sure about these, if they don't cause any errors - maybe we should turn them off?
- `testing-library/no-await-sync-query`
- `testing-library/no-dom-import`
- `testing-library/no-wait-for-empty-callback`

MichaelDeBoey replied:
For now I kept the `testing-library/no-await-sync-query`, `testing-library/no-dom-import` & `testing-library/no-wait-for-empty-callback` rules
Let me know if they need to be removed 🙂

mrmckeb reviewed (commented):
Thanks for the updates here @MichaelDeBoey - and for all of your work on this so far.

I've left some thoughts around other rules that we could remove. As previously mentioned, the team want to keep `eslint-config-react-app` very minimal (which was the original philosophy). 

If you think I've missed/misunderstood anything, let me know - I read the descriptions for each rule, but in some cases it was harder to tell if the rule was for potential errors/issues, or just good practice.

I'd love to also get some thoughts from @ianschmitz and @iansu here.

jugshaurya commented:
Now since we have added testing library eslint configs to cra, we don't need to add eslint-plugin-jest and eslint-plugin-testing-library to our project? Am I correct? Because i was going through @kentcdodds article, https://kentcdodds.com/blog/common-mistakes-with-react-testing-library , and found out these plugins as recommended.

Belco90 commented:
@jugshaurya you are right! Create React App already includes them since v4. [We still need to mention this on `eslint-plugin-testing-library` README to clarify it's included with CRA already](https://github.com/testing-library/eslint-plugin-testing-library/issues/242).
